### PR TITLE
feat: [SRM-12695]: Add recalculation for composite SLO

### PR DESCRIPTION
**Pull request by Varsha-lalwani**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/40036)

### CODE REVIEW

**Comment by Varsha-lalwani** [comment]
Retrigger featurename check

**Comment by Varsha-lalwani** [comment]
trigger featurenamecheck

File: 300-cv-nextgen/src/main/java/io/harness/cvng/statemachine/services/api/CompositeSLOMetricAnalysisStateExecutor.java [hunk 1]
```diff
@@ -118,6 +143,6 @@ public AnalysisState handleTransition(CompositeSLOMetricAnalysisState analysisSt
 
   @Override
   public AnalysisState handleRetry(CompositeSLOMetricAnalysisState analysisState) {
-    return analysisState;
+    return execute(analysisState);
```
Comment (naveen-harness) [inline 300-cv-nextgen/src/main/java/io/harness/cvng/statemachine/services/api/CompositeSLOMetricAnalysisStateExecutor.java:146]: If retryCount is max retry, analysisState should be in failed state.

File: 300-cv-nextgen/src/main/java/io/harness/cvng/statemachine/services/api/CompositeSLOMetricAnalysisStateExecutor.java [hunk 2]
```diff
@@ -10,6 +10,7 @@
 import io.harness.cvng.core.beans.params.ProjectParams;
```
Comment (naveen-harness) [inline 300-cv-nextgen/src/main/java/io/harness/cvng/statemachine/services/api/CompositeSLOMetricAnalysisStateExecutor.java:10]: We need to implement rerun and transition as well, please check. 
For rerun the analysis state would go into RUNNING state back again.

**Comment by diptiman-adak** [comment]
trigger checkstylecodeformat

**Review by deepak-harness — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40036)* status: 9 file(s) to review. 

___
*[`300-cv-nextgen/src/main/java/io/harness/cvng/VerificationApplication.java` line 973 at r2](https://reviewable.io/reviews/harness/harness-core/40036#-NH-PJ7o7FrZzOTx32ZS:-NH-PJ7pAjHk_52FtFIN:b59ld1o) ([raw file](https://github.com/harness/harness-core/blob/35e8b01b4fa6abec4c7bc913de5a311aad7c2e17/300-cv-nextgen/src/main/java/io/harness/cvng/VerificationApplication.java#L973)):*
> ```Java
>                 -> query.and(
>                     query.criteria(ServiceLevelObjectiveV2Keys.type).equal(ServiceLevelObjectiveType.COMPOSITE),
>                     query.criteria(ServiceLevelObjectiveV2Keys.createNextTaskIteration).equal(0L)))
> ```

nit: we should re-think if we can move the creation of initial state machine to create SLO method.
___
*[`300-cv-nextgen/src/main/java/io/harness/cvng/statemachine/services/impl/OrchestrationServiceImpl.java` line 77 at r2](https://reviewable.io/reviews/harness/harness-core/40036#-NH-QhZoALxD441h4_xA:-NH-QhZoALxD441h4_xB:b-fnlmjw) ([raw file](https://github.com/harness/harness-core/blob/35e8b01b4fa6abec4c7bc913de5a311aad7c2e17/300-cv-nextgen/src/main/java/io/harness/cvng/statemachine/services/impl/OrchestrationServiceImpl.java#L77)):*
> ```Java
> 
>   @Override
>   public void queueAnalysisOnce(String verificationTaskId, String accountId, Instant startTime, Instant endTime) {
> ```

nit: lets rename it to queueAnalysisWithoutEventPublish

**Review by Varsha-lalwani — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40036)* status: 9 file(s) to review. 2. unresolved discussions. 

___
*[`300-cv-nextgen/src/main/java/io/harness/cvng/statemachine/services/api/CompositeSLOMetricAnalysisStateExecutor.java` line 10 at r1](https://reviewable.io/reviews/harness/harness-core/40036#-NGbDbcAImV9abG6nzT0-r1-10:-NH080Rn4ChGk-8auG8-:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/003ad203241ca94ed3fd980eb09ea442b35c35ef/300-cv-nextgen/src/main/java/io/harness/cvng/statemachine/services/api/CompositeSLOMetricAnalysisStateExecutor.java#L10)):*
<details><summary><i>Previously, naveen-harness (Naveen Shukla) wrote…</i></summary><blockquote>

We need to implement rerun and transition as well, please check. 
For rerun the analysis state would go into RUNNING state back again.
</blockquote></details>

Done.
___
*[`300-cv-nextgen/src/main/java/io/harness/cvng/statemachine/services/api/CompositeSLOMetricAnalysisStateExecutor.java` line 146 at r1](https://reviewable.io/reviews/harness/harness-core/40036#-NGbDbcAImV9abG6nzT0-r1-146:-NH0817D33zVfRvCl_T4:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/003ad203241ca94ed3fd980eb09ea442b35c35ef/300-cv-nextgen/src/main/java/io/harness/cvng/statemachine/services/api/CompositeSLOMetricAnalysisStateExecutor.java#L146)):*
<details><summary><i>Previously, naveen-harness (Naveen Shukla) wrote…</i></summary><blockquote>

If retryCount is max retry, analysisState should be in failed state.
</blockquote></details>

Done.

File: 300-cv-nextgen/src/main/java/io/harness/cvng/statemachine/services/api/CompositeSLOMetricAnalysisStateExecutor.java [hunk 3]
```diff
@@ -45,15 +49,33 @@ public class CompositeSLOMetricAnalysisStateExecutor extends AnalysisStateExecut
...
+    CompositeSLORecord lastSLORecord = compositeSLORecordService.getLatestCompositeSLORecordWithVersion(
+        compositeServiceLevelObjective.getUuid(), compositeServiceLevelObjective.getVersion());
+    LocalDateTime currentLocalDate =
+        LocalDateTime.ofInstant(clock.instant(), compositeServiceLevelObjective.getZoneOffset());
+    Instant startTimeForCurrentRange = compositeServiceLevelObjective.getCurrentTimeRange(currentLocalDate)
+                                           .getStartTime(compositeServiceLevelObjective.getZoneOffset());
+    Instant startedAtTime = Instant.ofEpochMilli(compositeServiceLevelObjective.getStartedAt());
+    Instant startTime = (startTimeForCurrentRange.isAfter(startedAtTime)) ? startTimeForCurrentRange : startedAtTime;
+    if (lastSLORecord != null) {
+      startTime = Instant.ofEpochSecond(lastSLORecord.getEpochMinute() * 60 + 60);
```
Comment (naveen-harness) [inline 300-cv-nextgen/src/main/java/io/harness/cvng/statemachine/services/api/CompositeSLOMetricAnalysisStateExecutor.java:73]: We should keep + 60 as a constant which helps in understanding it's significance as well, something like START_TIME_OFFSET ?

**Review by naveen-harness — APPROVED** [review]

**Review by deepak-harness — APPROVED** [review]
> Reviewed 4 of 9 files at r1, 5 of 6 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40036)* status: 2. unresolved discussions.

**Review by Varsha-lalwani — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40036#-:-NH8Vnzf61WvGceHSW74:bnfp4nl)* status: 3. unresolved discussions. 

___
*[`300-cv-nextgen/src/main/java/io/harness/cvng/statemachine/services/api/CompositeSLOMetricAnalysisStateExecutor.java` line 73 at r3](https://reviewable.io/reviews/harness/harness-core/40036#-NGbDbcAImV9abG6nzT0-r3-73:-NH8VmCq06ZFu5-m6qib:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/157ddea9bad3d75ed0919bf57252f866e601bc60/300-cv-nextgen/src/main/java/io/harness/cvng/statemachine/services/api/CompositeSLOMetricAnalysisStateExecutor.java#L73)):*
<details><summary><i>Previously, naveen-harness (Naveen Shukla) wrote…</i></summary><blockquote>

We should keep + 60 as a constant which helps in understanding it's significance as well, something like START_TIME_OFFSET ? 
</blockquote></details>

Done.

**Pull request merged**